File: OJ/usaco/s1/s14/milk3.pas
{
ID:lijianx1
PROG:milk3
LANG:PASCAL
}
program milk3;
var a,b,c,i:longint;
    s:array[0..20,0..20] of boolean;
    first:boolean;
function min(a,b:longint):longint;
 begin
     if a<b then min:=a else min:=b;
 end;

procedure bfs(ra,rc:longint);
 var rb:longint;
 begin
     if s[ra,rc] then exit;
     s[ra,rc]:=true;
     rb:=c-ra-rc;
     bfs(ra-min(ra,b-rb),rc); {a-->b}
     bfs(ra-min(ra,c-rc),rc+min(ra,c-rc));{a-->c};
     bfs(ra+min(rb,a-ra),rc);{b-->a}
     bfs(ra,rc+min(rb,c-rc));{b-->c}
     bfs(ra+min(rc,a-ra),rc-min(rc,a-ra));{c-->a}
     bfs(ra,rc-min(rc,b-rb));{c-->b}
 end;

begin
    assign(input,'milk3.in'); reset(input);
    assign(output,'milk3.out'); rewrite(output);
    readln(a,b,c);
    fillchar(s,sizeof(s),false);
    bfs(0,c);
    first:=true;
    for i:=0 to c do
       if s[0,i] then begin
           if first then begin
              write(i);
              first:=false;
           end else write(' ',i);
       end;
    writeln;
    close(input);
    close(output);
end.

File: OJ/usaco/cpp/milk3.cpp
/*
ID:oifox20071
LANG:C++
PROG:milk3
*/
#include <iostream>
using namespace std;
const int limitsize = 21;
bool flag[limitsize][limitsize] = {0};
int a, b, c;
void bfs(int ra, int rc)
{
	if (flag[ra][rc]) return;
	flag[ra][rc] =  1;
	int rb = c - ra - rc;
	bfs(ra - min(ra, b - rb), rc);
	bfs(ra - min(ra, c - rc), rc + min(ra, c - rc));
	bfs(ra + min(rb, a - ra), rc);
	bfs(ra, rc + min(rb, c - rc));
	bfs(ra + min(rc, a - ra), rc - min(rc, a - ra));
	bfs(ra, rc - min(rc, b - rb));
}

int main()
{
	freopen("milk3.in", "r", stdin); freopen("milk3.out", "w", stdout);
	cin >> a >> b >> c;
	bfs(0, c);
	bool first = true;
	for (int i = 0; i <= c; ++ i)
		if (flag[0][i])
		{
			if (first)
			{
				cout << i;
				first = 0;
			}
			else			
				cout << " " << i;
		}
	cout << endl;
	return 0;
}
